stripped/colabs/Implementacja_NearestNeighbours_stripped.py

```diff
@@ -0,0 +1,177 @@
+    .withColumn('Month', F.col('Date') % 100)
+    .drop('Date')
+)
+nasa.show(5)
+
+
+nasa_coords = spark.sql("""SELECT DISTINCT lat, lon FROM nasa""")
+nasa_coords.collect()
+
+
+"""
+### **2. Implementacja algorytmu kNN:**
+"""
+
+# function searches for points that lie within a (euclidean) ball of size *radius* around the query points
+# if optional argument *k* is given then function searches for (at most) k nearest points that lie within
+# a (euclidean) ball of size *radius* around the query points
+
+#point = (latitude, longitude)
+def kRadiusNN(df: SparkDataFrame, radius: float, point: Tuple[float, float], k: Optional[int]=None) -> pd.DataFrame :
+  assert (25.0625<= point[0] <=52.9375) and (-124.9375<= point[1] <=-67.0625 ), 'Wrong coordinates (out of range)'
+
+  df = df.collect()
+  neighbours_pd = pd.DataFrame({"lon":[], "lat":[], 'dist':[]})
+
+  for row in df:
+
+     lon = row['lon']
+     lat = row['lat']
+     dist = geodesic((lat, lon), point).km
+     if dist <= radius:
+        new_row = {'lon': lon , 'lat': lat, 'dist' : dist}
+        neighbours_pd.loc[len(neighbours_pd)] = new_row
+
+  if k and (k < len(neighbours_pd)):
+
+      neighbours_pd = neighbours_pd.sort_values('dist', ascending=True)
+      if not (neighbours_pd.at[k-1,'dist']==neighbours_pd.at[k,'dist']):  #checking if there is no tie (more neighbours with the same distance)
+         neighbours_pd = neighbours_pd.iloc[:k]
+      else:
+         raise Exception("Unable to determine k nearest neighbours: more neighbours with the same distance")
+
+  return(neighbours_pd)
+
+
+kRadiusNN(nasa_coords, 20, (35, -110), 5)
+
+
+# weighted: if True then function will weight points by the inverse of their distance (in this case, closer neighbours of
+# a query point will have a greater influence than neighbors which are further away).
+
+def predict_class(df: SparkDataFrame, point: Tuple[float, float], radius: float, label_column_name: str,
+                  k: Optional[int]=None, weighted: Optional[bool]=False) -> bool :
+ if k:
+    neighbours = kRadiusNN(df, radius, point, k)
+ else:
+    neighbours = kRadiusNN(df, radius, point)
+
+ neighbours_rows = df.withColumn('neighbour', F.when( reduce (lambda cond1, cond2: (cond1) | (cond2),
+  [(F.col('lat') == neighbours.at[id, 'lat'])&(F.col('lon') == neighbours.at[id, 'lon'])
+  for id in neighbours.index]), 1).otherwise(0))
+
+ neighbours_rows = neighbours_rows.filter('neighbour = 1').toPandas()
+ neighbours_dist = pd.concat([neighbours_rows, neighbours], axis=1, join="inner")
+
+
+ if weighted:   #weighted nearest neighbours
+    neighbours_dist['dist'] = neighbours_dist['dist'].map(lambda x: 1/x)
+
+    label0 = neighbours_dist[neighbours_dist[label_column_name]==0]
+    label1 = neighbours_dist[neighbours_dist[label_column_name]==1]
+
+    frequency0 = label0['dist'].sum()
+    frequency1 = label1['dist'].sum()
+
+    return 0 if frequency0 > frequency1 else 1
```

Comment:
Nie próbujmy na siłę wyprzedzać zadanie. Zrobienie od razu też wersji ważonej znacząco komplikuje interfejs a nie wiadomo czy się przyda. Zadania starajmy się rozwiązywać minimalnym rozbudowaniem.

Reply:
Jak wyżej: wyjaśniam w korespondencji na MS Teams.

Reply:
Starajmy się nie przenosić komunikacji do innych kanałów.

Reply:
Rozumiem, po prostu myślalam, że komentarze na githubie, to nie miejsce na dłuższe wypowiedzi....